// ocache/dist/index.mjs
import { hash } from "ohash";
//#region src/storage.ts
/** Creates an in-memory storage backed by a `Map` with optional TTL support (in seconds). */
function createMemoryStorage() {
	const map = /* @__PURE__ */ new Map();
	return {
		get(key) {
			const entry = map.get(key);
			if (!entry) return null;
			if (entry.expires && Date.now() > entry.expires) {
				map.delete(key);
				return null;
			}
			return entry.value;
		},
		set(key, value, opts) {
			map.set(key, {
				value,
				expires: opts?.ttl ? Date.now() + opts.ttl * 1e3 : void 0
			});
		}
	};
}
let _storage;
/** Returns the current storage instance. If none has been set via `setStorage`, lazily initializes an in-memory storage. */
function useStorage() {
	if (!_storage) _storage = createMemoryStorage();
	return _storage;
}
/** Sets a custom storage implementation to be used by all cached functions. */
function setStorage(storage) {
	_storage = storage;
}
//#endregion
//#region src/cache.ts
function defaultCacheOptions$1() {
	return {
		name: "_",
		base: "/cache",
		swr: true,
		maxAge: 1
	};
}
/**
* Wraps a function with caching support including TTL, SWR, integrity checks, and request deduplication.
*
* @param fn - The function to cache.
* @param opts - Cache configuration options.
* @returns A new async function that returns cached results when available.
*/
function defineCachedFunction(fn, opts = {}) {
	opts = {
		...defaultCacheOptions$1(),
		...opts
	};
	const pending = {};
	const group = opts.group || "ocache/functions";
	const name = opts.name || fn.name || "_";
	const integrity = opts.integrity || hash([fn, opts]);
	const validate = opts.validate || ((entry) => entry.value !== void 0);
	const _onError = (context, error) => {
		if (opts.onError) opts.onError(error);
		else console.error(context, error);
	};
	async function get(key, resolver, shouldInvalidateCache, event) {
		const cacheKey = [
			opts.base,
			group,
			name,
			key + ".json"
		].filter(Boolean).join(":").replace(/:\/$/, ":index");
		let entry = await Promise.resolve(useStorage().get(cacheKey)).catch((error) => {
			_onError("[cache] Cache read error.", error);
		}) || {};
		if (typeof entry !== "object") {
			entry = {};
			_onError("[cache]", /* @__PURE__ */ new Error("Malformed data read from cache."));
		}
		const ttl = (opts.maxAge ?? 0) * 1e3;
		if (ttl) entry.expires = Date.now() + ttl;
		const expired = shouldInvalidateCache || entry.integrity !== integrity || ttl && Date.now() - (entry.mtime || 0) > ttl || validate(entry) === false;
		const _resolve = async () => {
			const isPending = pending[key];
			if (!isPending) {
				if (entry.value !== void 0 && (opts.staleMaxAge || 0) >= 0 && opts.swr === false) {
					entry.value = void 0;
					entry.integrity = void 0;
					entry.mtime = void 0;
					entry.expires = void 0;
				}
				pending[key] = Promise.resolve(resolver());
			}
			try {
				entry.value = await pending[key];
			} catch (error) {
				if (!isPending) delete pending[key];
				throw error;
			}
			if (!isPending) {
				entry.mtime = Date.now();
				entry.integrity = integrity;
				delete pending[key];
				if (validate(entry) !== false) {
					let setOpts;
					if (opts.maxAge && !opts.swr) setOpts = { ttl: opts.maxAge };
					const promise = Promise.resolve(useStorage().set(cacheKey, entry, setOpts)).catch((error) => {
						_onError("[cache] Cache write error.", error);
					});
					if ((event?.req)?.waitUntil) event.req.waitUntil(promise);
				}
			}
		};
		const _resolvePromise = expired ? _resolve() : Promise.resolve();
		if (entry.value === void 0) await _resolvePromise;
		else if (expired && (event?.req)?.waitUntil) event.req.waitUntil(_resolvePromise);
		if (opts.swr && validate(entry) !== false) {
			_resolvePromise.catch((error) => {
				_onError("[cache] SWR handler error.", error);
			});
			return entry;
		}
		return _resolvePromise.then(() => entry);
	}
	return async (...args) => {
		if (await opts.shouldBypassCache?.(...args)) return fn(...args);
		const entry = await get(await (opts.getKey || getKey)(...args), () => fn(...args), await opts.shouldInvalidateCache?.(...args), isHTTPEvent(args[0]) ? args[0] : void 0);
		let value = entry.value;
		if (opts.transform) value = await opts.transform(entry, ...args) || value;
		return value;
	};
}
/** Alias for {@link defineCachedFunction}. */
const cachedFunction = defineCachedFunction;
function isHTTPEvent(input) {
	return input?.req instanceof Request;
}
function getKey(...args) {
	return args.length > 0 ? hash(args) : "";
}
//#endregion
//#region src/http.ts
function defaultCacheOptions() {
	return {
		name: "_",
		base: "/cache",
		swr: true,
		maxAge: 1
	};
}
/**
* Wraps an HTTP event handler with response caching.
*
* Automatically generates cache keys from the URL path and variable headers,
* sets `cache-control`, `etag`, and `last-modified` headers, and handles
* `304 Not Modified` responses via conditional request headers.
*
* @param handler - The event handler to cache.
* @param opts - Cache and HTTP-specific configuration options.
* @returns A new event handler that serves cached responses when available.
*/
function defineCachedHandler(handler, opts = defaultCacheOptions()) {
	const variableHeaderNames = (opts.varies || []).filter(Boolean).map((h) => h.toLowerCase()).sort();
	const _toResponse = opts.toResponse || ((rawValue) => rawValue instanceof Response ? rawValue : new Response(String(rawValue)));
	const _createResponse = opts.createResponse || ((body, init) => new Response(body, init));
	const _handleCacheHeaders = opts.handleCacheHeaders || _defaultHandleCacheHeaders;
	const _cachedHandler = cachedFunction(async (event) => {
		const filteredHeaders = [...event.req.headers.entries()].filter(([key]) => !variableHeaderNames.includes(key.toLowerCase()));
		try {
			const originalReq = event.req;
			event.req = new Request(event.req.url, {
				method: event.req.method,
				headers: filteredHeaders
			});
			if (originalReq.runtime) event.req.runtime = originalReq.runtime;
		} catch (error) {
			console.error("[cache] Failed to filter headers:", error);
		}
		const res = await _toResponse(await handler(event), event);
		const body = await res.text();
		if (!res.headers.has("etag")) res.headers.set("etag", `W/"${hash(body)}"`);
		if (!res.headers.has("last-modified")) res.headers.set("last-modified", (/* @__PURE__ */ new Date()).toUTCString());
		const cacheControl = [];
		if (opts.swr) {
			if (opts.maxAge) cacheControl.push(`s-maxage=${opts.maxAge}`);
			if (opts.staleMaxAge) cacheControl.push(`stale-while-revalidate=${opts.staleMaxAge}`);
			else cacheControl.push("stale-while-revalidate");
		} else if (opts.maxAge) cacheControl.push(`max-age=${opts.maxAge}`);
		if (cacheControl.length > 0) res.headers.set("cache-control", cacheControl.join(", "));
		return {
			status: res.status,
			statusText: res.statusText,
			headers: Object.fromEntries(res.headers.entries()),
			body
		};
	}, {
		...opts,
		shouldBypassCache: (event) => {
			return event.req.method !== "GET" && event.req.method !== "HEAD";
		},
		getKey: async (event) => {
			const customKey = await opts.getKey?.(event);
			if (customKey) return escapeKey(customKey);
			const _url = event.url ?? new URL(event.req.url);
			const _path = _url.pathname + _url.search;
			let _pathname;
			try {
				_pathname = escapeKey(decodeURI(new URL(_path, "http://localhost").pathname)).slice(0, 16) || "index";
			} catch {
				_pathname = "-";
			}
			return [`${_pathname}.${hash(_path)}`, ...variableHeaderNames.map((header) => [header, event.req.headers.get(header)]).map(([name, value]) => `${escapeKey(name)}.${hash(value)}`)].join(":");
		},
		validate: (entry) => {
			if (!entry.value) return false;
			if (entry.value.status >= 400) return false;
			if (entry.value.body === void 0) return false;
			if (entry.value.headers.etag === "undefined" || entry.value.headers["last-modified"] === "undefined") return false;
			return true;
		},
		group: opts.group || "cache/handlers",
		integrity: opts.integrity || hash([handler, opts])
	});
	return async (event) => {
		if (opts.headersOnly) {
			if (_handleCacheHeaders(event, { maxAge: opts.maxAge })) return _createResponse(null, { status: 304 });
			return handler(event);
		}
		const response = await _cachedHandler(event);
		if (_handleCacheHeaders(event, {
			modifiedTime: new Date(response.headers["last-modified"]),
			etag: response.headers.etag,
			maxAge: opts.maxAge
		})) return _createResponse(null, { status: 304 });
		return _createResponse(response.body ?? null, {
			status: response.status,
			statusText: response.statusText,
			headers: response.headers
		});
	};
}
function escapeKey(key) {
	return String(key).replace(/\W/g, "");
}
function _defaultHandleCacheHeaders(event, conditions) {
	const ifNoneMatch = event.req.headers.get("if-none-match");
	if (ifNoneMatch && conditions.etag && ifNoneMatch === conditions.etag) return true;
	const ifModifiedSince = event.req.headers.get("if-modified-since");
	if (ifModifiedSince && conditions.modifiedTime) {
		if (new Date(ifModifiedSince) >= conditions.modifiedTime) return true;
	}
	return false;
}
//#endregion
export { cachedFunction, createMemoryStorage, defineCachedFunction, defineCachedHandler, setStorage, useStorage };
