// env-runner/dist/_chunks/runner.mjs
import { t as BaseEnvRunner } from "./base-runner.mjs";
import { existsSync } from "node:fs";
import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";
//#region src/runners/deno-process/runner.ts
let _defaultEntry;
var DenoProcessEnvRunner = class extends BaseEnvRunner {
	#process;
	constructor(opts) {
		_defaultEntry ||= fileURLToPath(import.meta.resolve("env-runner/runners/deno-process/worker"));
		super({
			...opts,
			workerEntry: opts.workerEntry || _defaultEntry
		});
		this.#initProcess(opts.execArgv);
	}
	sendMessage(message) {
		if (!this.#process) throw new Error("Deno env process should be initialized before sending messages.");
		this.#process.send(message);
	}
	_hasRuntime() {
		return Boolean(this.#process);
	}
	_runtimeType() {
		return "process";
	}
	async _closeRuntime() {
		if (!this.#process) return;
		this.#process.removeAllListeners?.();
		try {
			this.#process.kill();
		} catch {}
		this.#process = void 0;
	}
	#initProcess(execArgv) {
		if (!existsSync(this._workerEntry)) {
			this.close(`process entry not found in "${this._workerEntry}".`);
			return;
		}
		const env = {
			...process.env,
			ENV_RUNNER_NAME: this._name,
			ENV_RUNNER_DATA: JSON.stringify(this._data || {})
		};
		const child = spawn("deno", [
			"run",
			"--allow-all",
			"--node-modules-dir=auto",
			...execArgv || [],
			this._workerEntry
		], {
			env,
			stdio: [
				"pipe",
				"pipe",
				"pipe",
				"ipc"
			]
		});
		const exited = new Promise((resolve) => {
			child.once("exit", (code) => resolve(code ?? 1));
		});
		const handle = {
			pid: child.pid,
			kill: () => child.kill(),
			send: (message) => child.send(message),
			exited,
			_exitCode: void 0,
			removeAllListeners: () => child.removeAllListeners()
		};
		child.once("exit", (code) => {
			handle._exitCode = code;
			this.close(`process exited with code ${code}`);
		});
		child.on("error", (error) => {
			if (!this.closed) {
				console.error(`Process error:`, error);
				this.close(error);
			}
		});
		child.on("message", (message) => {
			this._handleMessage(message);
		});
		this.#process = handle;
	}
};
//#endregion
export { DenoProcessEnvRunner as t };
